sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsSpanImpl.java

```diff
@@ -0,0 +1,743 @@
+import io.opentelemetry.proto.trace.v1.TruncatableString;
+import io.opentelemetry.resource.Resource;
+import io.opentelemetry.sdk.internal.Clock;
+import io.opentelemetry.sdk.internal.TimestampConverter;
+import io.opentelemetry.sdk.trace.config.TraceConfig;
+import io.opentelemetry.sdk.trace.internal.ConcurrentIntrusiveList.Element;
+import io.opentelemetry.trace.AttributeValue;
+import io.opentelemetry.trace.Event;
+import io.opentelemetry.trace.Link;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.SpanContext;
+import io.opentelemetry.trace.SpanData;
+import io.opentelemetry.trace.SpanData.TimedEvent;
+import io.opentelemetry.trace.SpanId;
+import io.opentelemetry.trace.Status;
+import io.opentelemetry.trace.TraceId;
+import io.opentelemetry.trace.Tracer;
+import io.opentelemetry.trace.Tracestate;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+/** Implementation for the {@link Span} class that records trace events. */
+@ThreadSafe
+public final class RecordEventsSpanImpl implements SpanSdk, Element<RecordEventsSpanImpl> {
+  private static final Logger logger = Logger.getLogger(Tracer.class.getName());
+
+  // Contains the identifiers associated with this Span.
+  private final SpanContext context;
+  // The parent SpanId of this span. Null if this is a root span.
+  @Nullable private final SpanId parentSpanId;
+  // True if the parent is on a different process.
+  @Nullable private final Boolean hasRemoteParent;
+  // Active trace configs when the Span was created.
+  private final TraceConfig traceConfig;
+  // Handler called when the span starts and ends.
+  private final StartEndHandler startEndHandler;
+  // The displayed name of the span.
+  @GuardedBy("this")
+  private String name;
+  // The kind of the span.
+  @Nullable private final Kind kind;
+  // The clock used to get the time.
+  private final Clock clock;
+  // The time converter used to convert nano time to Timestamp. This is needed because Java has
+  // millisecond granularity for Timestamp and tracing events are recorded more often.
+  private final TimestampConverter timestampConverter;
+  // The resource associated with this span.
+  private final Resource resource;
+  // The start time of the span.
+  private final long startNanoTime;
+  // Set of recorded attributes. DO NOT CALL any other method that changes the ordering of events.
+  @GuardedBy("this")
+  @Nullable
+  private AttributesWithCapacity attributes;
+  // List of recorded events.
+  @GuardedBy("this")
+  @Nullable
+  private TraceEvents<TimedEvent> events;
+  // List of recorded links to parent and child spans.
+  @GuardedBy("this")
+  @Nullable
+  private TraceEvents<Link> links;
```

Comment:
Is this TraceEvents<Links>???

Reply:
`TraceEvents` is just an internal helper class used to track the dropped count. This is copied from https://github.com/census-instrumentation/opencensus-java/blob/282d742f64c4ffd2273608d644bffbefe538afdd/impl_core/src/main/java/io/opencensus/implcore/trace/RecordEventsSpanImpl.java#L92-L95.